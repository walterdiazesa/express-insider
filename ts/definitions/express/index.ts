export enum HandlerType {
  ROUTE = "Route",
  MIDDLEWARE = "Middleware",
  ROUTE_MIDDLEWARE = "Route Middleware",
}

export * from './objects';
export * from './verbs';