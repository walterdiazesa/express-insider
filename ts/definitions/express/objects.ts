import { TrailOptions } from "..";
import { HandlerType, Method } from "../../../ts";

export type RouteMatcher = {
  route: TrailOptions['ignoreRoutes'][number]['route'],
  method: Lowercase<TrailOptions['ignoreRoutes'][number]['method']>,
};

export type Route = {
  path: `/${string}`;
  stack: RouteStack[];
  methods: Record<Method, boolean>;
};

type Stack<T extends HandlerType | undefined = undefined> = {
  handle: Function & { stack?: (StackItem | StackItem<HandlerType.ROUTE>)[] };
  name: string;
  params?: undefined;
  path: T extends HandlerType.ROUTE ? string : unknown;
  keys: Record<string, string | boolean | number>[];
  regexp: RegExp;
};

export type StackItem<T extends HandlerType | undefined = undefined> = Stack<T> & {
  route: T extends undefined ? undefined | Route : T extends HandlerType.ROUTE ? Route : undefined;
  // Config params
  ignore: T extends HandlerType.ROUTE ? undefined | true : never;
  showResponse: T extends HandlerType.ROUTE ? undefined | true : never;
  showRequestedURL: T extends HandlerType.ROUTE ? undefined | true : never;
}

export type RouteStack<T extends Method = Method> = Stack & {
  method: T;
};