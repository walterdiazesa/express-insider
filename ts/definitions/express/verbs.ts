export type Method = "get" | "post" | "options" | "all" | "patch" | "delete" | "put" | "head" | "connect" | "trace";