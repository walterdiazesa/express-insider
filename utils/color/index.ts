import { COLOR } from "../../constants";

export const getStatusCodeColor = (statusCode: number) => (statusCode >= 200 && statusCode <= 299 ? COLOR.fgGreen : statusCode >= 400 ? COLOR.fgRed : COLOR.fgWhite);