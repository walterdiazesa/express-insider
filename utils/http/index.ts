import { Response } from "express";
import { HTTP_STATUS_MESSAGE_TO_CODE } from "../../constants";

/**
  * __[ODD-1]__: When using express@4.9.0 (unknown about other versions under 4.10.0) the test "trail on custom configuration â€º Should handle [CASE 12] with POST /conditional-route"
  * fails, even that the route return statusCode 200 as expected, the trail output shows 404, when removing the next call in the third route everything works as expected,
  * what is happenning is that express treats undefined next routes different in 4.9.0, so as no afterwards middleware is found the app crashes setting the statusCode as 404, so
  * the bandage fix to support 4.9.0 is to redefine the possible outcome for 404 cases if res.statusMessage is still defined as "OK"
  * 
  * Current state: FIXED!, we could return => res.statusCode directly, but it's a good redundant condition
  */
export const getStatusCode = (res: Response): typeof HTTP_STATUS_MESSAGE_TO_CODE[keyof typeof HTTP_STATUS_MESSAGE_TO_CODE] =>
  res.statusCode === 404 && res.statusMessage !== 'Not found' ? HTTP_STATUS_MESSAGE_TO_CODE[res.statusMessage.toLowerCase().replaceAll(' ', '')] : res.statusCode;