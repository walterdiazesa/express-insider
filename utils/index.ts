import { Express } from 'express';
import { getCfg } from '../config';
import { HandlerType, Route, RouteMatcher, StackItem } from '../ts';
import { generateRouteMatcherGroup } from './config';

const config = getCfg();

export const getRuntime = () => {
  const runtime = (process as any).isBun || process.versions.bun ? 'bun' : 'node'
  const { engines } = require('../package.json');
  return {
    runtime,
    version: process.versions[runtime],
    trailCompatibility: engines[runtime]
  }
}

const sanitizePackageVersion = (version: string) => version.match(/[0-9.]+/).shift() as `${number}.${number}.${number}`;

export const isRuntimeCompatible = (runtime: ReturnType<typeof getRuntime>) => sanitizePackageVersion(runtime.version).localeCompare(sanitizePackageVersion(runtime.trailCompatibility), undefined, { numeric: true, sensitivity: 'base' }) >= 0

export const getRSS = (): '' | ` [${number} bytes]` => {
  if (!config[6]) return ''
  // BUN: memoryUsage on process since v0.6.14
  let rss = process.memoryUsage?.().rss;
  /* istanbul ignore next */ 
  if (rss === undefined) {
    try {
      // BUN: bun:jsc available since v0.5.1
      const { memoryUsage } = require('bun:jsc');
      rss = memoryUsage().current
    } catch (e) {}
  }
  return ` [${rss} bytes]`
}

export const isStackItemRoute = (stackItem: StackItem | StackItem<HandlerType.ROUTE>): stackItem is StackItem<HandlerType.ROUTE> => typeof stackItem.route === 'object';

export const formatAnonymousRoute = (idx: number): `<anonymous (${number})>` => `<anonymous (${idx})>`;

/**
 * __[ODD-3]__: Under 4.0 ```app.stack``` is defined, after is ```app._router.stack```
 */
export const getStack = (app: Express): (StackItem | StackItem<HandlerType.ROUTE>)[] => app.stack || app._router.stack

/*
  PAPER: Let's start by saying that the timing difference between the worst possible matcher (a simple naive implementation
  without caching or transforming the original RouteMatcher given by the consumer) and this approach of grouping and storing
  the RouteMatchers in pairs inside an array is insignificant, not only in real-case scenarios but also relatively speaking.

  However, if I can either mantain the same performance (in the worst case) or improve it just a little bit without causing
  any issues in the ecosystem (the app itself or the debugging process for other developers, though in this case, I'm the
  only one :p), why not?

  My first approach was caching results by grouping a Map inside a Map (for caching, this was the faster implementation). I
  also tested objects, storing keys instead of references, and using the route as the identifier for the outer Map and the
  matchGroup for the inner Map (and vice versa), etc.
  
  But there were many variables that significantly affected the performance results:
    - The order of the route matcher requested versus the route that is actually being requested
      (User requesting GET / --> Match:
          [{ route: '/', method: 'post' }, { route: '/todo', method: 'get' }, { route: '/', method: 'get' }])
          is different than
          [{ route: '/', method: 'get' }, { route: '/todo', method: 'get' }, { route: '/', method: 'post' }])
    - And even worst, the size of the RouteMatcher set, at which point exactly is caching worth it? 3 items? 9 items? 20 items?
      keep in mind that the previous constraint (The order of the route matcher) is also affecting the second point both
      separately and as a whole
          Naive faster: [{ route: '/', method: 'post' }, { route: '/todo', method: 'get' }, { route: '/', method: 'get' }]
          Map faster?: [{ route: '/', method: 'post' }, (3 items more...), { route: '/', method: 'get' }]

  Then, I thought of transforming the original RouteMatcher into an array because, as I already mentioned in Config, using
  arrays is faster than any other type of object (even though arrays are also "objects").

  So, I came up with an array of arrays ([['/', ['get', 'post']], ['/todo', ['put']]]) which worked better than the previous
  approaches in all the cases.
  
  But then I thought, "Mmmmm, maybe using a top-level array of pairs with only the methods encapsuled in an array is faster."
  Thus, instead of [['/', ['get', 'post']], ['/todo', ['put']]]
  I used ['/', ['get', 'post'], '/todo', ['put']] and my hypothesis was correct â€“ which is the approach you're seeing now.
*/
export const isRouteMatching = (route: Route, matchGroup: ReturnType<typeof generateRouteMatcherGroup>) => {
  if (!matchGroup || typeof matchGroup === 'boolean') return !!matchGroup;
  const path = route.path;
  for (let i = 0; i < matchGroup.length; i+=2) {
    const matchItemPath = matchGroup[i] as Route['path'];
    if (matchItemPath !== path) continue;
    const matchItemMethods = matchGroup[i + 1] as RouteMatcher['method'][];
    for (let j = 0; j < matchItemMethods.length; j++) {
      const method = matchItemMethods[j];
      if (method === 'any' || route.methods[method]) {
        return true;
      }
    }
  }
  return false
}

export * from './report'
export * from './json'
export * from './color'
export * from './logger'
export * from './http'
export * from './config'