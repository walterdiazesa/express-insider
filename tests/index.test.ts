import request from 'supertest'
import express from 'express'
import bodyparser from 'body-parser'
import { trail } from '../index';
import { CASE_2_SEGMENT_FETCH_SNAPSHOT, CASE_2_SEGMENT_NO_FETCH_SNAPSHOT } from './__custom_snapshots__';
import { awaitConsoleLog, sleep } from './utils';

const isFetchAvailable = () => typeof fetch === 'function'; // parseFloat(process.versions.node) >= 17.5

const app = express();

app.use(bodyparser.json()); // express.json() is only available since 4.16.0

app.get('/', (req, res) => {
  res.send({ working: true })
})

app.get('/no-response-route', (req, res) => {
  res.send({ payload: 'hidden' })
})

app.get('/ignore-route', (req, res) => {
  res.send({ working: false })
})

app.get('/response-route', (req, res) => {
  res.send({ payload: 'some payload', items: [{ id: 1 }, { id: 2 }] })
})

app.post(
  "/conditional-route",
  async function (req, res, next) {
    if (req.body.segment) {
      if (isFetchAvailable()) await req.logSegmentPerf("async fetch", async () => await fetch("https://jsonplaceholder.typicode.com/todos/1").then((page) => page.json()));
      await req.logSegmentPerf("async promise", () => sleep(100));
      await req.logSegmentPerf("async function", async () => await sleep(100, { payload: "some string payload" }));
      req.logSegmentPerf("sync", () => {
        console.log("sync perf");
        return { response: 99 };
      });
      const cb = req.logSegmentPerf("no cb");
      cb();
    }
    if (req.body.first) res.send({ first: true });
    if (req.body.firstCleaner) await sleep(100);
    if (req.body.next || req.body.second || req.body.third) next();
  },
  async function (req, res, next) {
    if (req.body.sleepBefore) await sleep(50);
    if (req.body.second) res.send({ second: true });
    if (req.body.sleepAfter) await sleep(150);
    if (req.body.next || req.body.third) next();
  },
  async function lastmiddle(req, res, next) {
    await sleep(150);
    if (req.body.third) res.send({ third: true });
    next();
  },
);

const reportFn = jest.fn();

trail(app, {
  trailId: () => 'test-id',
  timingFormatter: (elapsed) => 0,
  ignoreRoutes: [{ method: "any", route: "/ignore-route" }], 
  report: (id, payload) => {
    reportFn(id, payload)
  },
  showResponse: [{ method: "get", route: "/response-route" }],
});

describe('trail on custom configuration', () => {
  it('Should ignore GET /ignore-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).get("/ignore-route");
    expect(reportFn).toBeCalledTimes(0);
    expect(log.mock.calls.length).toBe(0);
  })
  it('Should NOT output GET /no-response-route payload', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).get("/no-response-route");
    await awaitConsoleLog(0, log.mock.calls)
    expect(log.mock.calls.toString()).not.toContain('[RESPONSE]')
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should output GET /response-route payload', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).get("/response-route");
    await awaitConsoleLog(0, log.mock.calls)
    expect(reportFn).toBeCalledTimes(7);
    expect(reportFn.mock.calls[0]).toEqual(["test-id", {"action": "start", "method": "GET", "reqUrl": "/response-route", "type": "wrapper"}])
    expect(reportFn.mock.calls[1]).toEqual(["test-id", {"elapsed": 0, "handlerName": "query", "isRouteHandler": false, "method": "GET", "type": "handler"}])
    expect(reportFn.mock.calls[2]).toEqual(["test-id", {"elapsed": 0, "handlerName": "expressInit", "isRouteHandler": false, "method": "GET", "type": "handler"}])
    expect(reportFn.mock.calls[3]).toEqual(["test-id", {"elapsed": 0, "handlerName": "jsonParser", "isRouteHandler": false, "method": "GET", "type": "handler"}])
    expect(reportFn.mock.calls[4]).toEqual(["test-id", {"elapsed": 0, "handlerName": "<anonymous (0)>", "isRouteHandler": true, "method": "GET", "reqUrl": "/response-route", "routeHandlerStage": "UNIQUE HANDLER", "statusCode": 200, "type": "handler"}])
    expect(reportFn.mock.calls[5]).toEqual(["test-id", {"trailId": "test-id", "method": "GET", "payload": "{\"payload\":\"some payload\",\"items\":[{\"id\":1},{\"id\":2}]}", "reqUrl": "/response-route", "routeHandlerStage": "UNIQUE HANDLER", "type": "report"}])
    expect(reportFn.mock.calls[6]).toEqual(["test-id", {"action": "finish", "elapsed": 0, "method": "GET", "reqUrl": "/response-route", "type": "wrapper"}])
    expect(log.mock.calls.toString()).toContain('[RESPONSE]')
    expect(log.mock.calls).toMatchSnapshot();
  })

  /**
   * CASE 1:
   * body: { first: true }
   * Should send response on first route middleware and not continue further
   * 
   * CASE 2:
   * body: { first: true, segment: true }
   * Should send response on first route middleware and not continue further while printing logSegmentPerf cases
   * 
   * CASE 3:
   * body: { first: true, firstCleaner: true }
   * Should send response on first route middleware and not continue further with a sleep function before finalizing
   * 
   * CASE 4:
   * body: { first: true, next: true }
   * Should send response on first route middleware and continue further
   * 
   * CASE 5:
   * body: { first: true, next: true, firstCleaner: true }
   * Should send response on first route middleware and continue further with a sleep function before finalizing
   * 
   * CASE 6:
   * body: { second: true }
   * Should send response on second route middleware
   * 
   * CASE 7:
   * body: { second: true, sleepBefore: true }
   * Should send response on second route middleware without cleanup nor total message
   * 
   * CASE 8:
   * body: { second: true, sleepAfter: true }
   * Should send response on second route middleware with cleanup and total message
   * 
   * CASE 9:
   * body: { second: true, sleepBefore: true, sleepAfter: true }
   * Should send response on second route middleware with correct route middlewares order
   * 
   * CASE 10:
   * body: { second: true, sleepBefore: true, sleepAfter: true, next: true }
   * Should send response on second route middleware with correct route middlewares order and continue
   * 
   * CASE 11:
   * body: { second: true, next: true }
   * Should send response on second route middleware and continue, check that is incorrectly ordered
   * 
   * CASE 12:
   * body: { third: true }
   * Should send response on third route middleware
  */
  it('Should handle [CASE 1] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ first: true });
    await awaitConsoleLog(0, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 2] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ first: true, segment: true });
    await awaitConsoleLog(0, log.mock.calls)
    expect(log.mock.calls.toString()).toBe(isFetchAvailable() ? CASE_2_SEGMENT_FETCH_SNAPSHOT : CASE_2_SEGMENT_NO_FETCH_SNAPSHOT)
  })
  it('Should handle [CASE 3] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ first: true, firstCleaner: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 4] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ first: true, next: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 5] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ first: true, next: true, firstCleaner: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 6] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ second: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 7] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ second: true, sleepBefore: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 8] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ second: true, sleepAfter: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 9] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ second: true, sleepBefore: true, sleepAfter: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 10] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ second: true, sleepBefore: true, sleepAfter: true, next: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 11] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ second: true, next: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
  it('Should handle [CASE 12] with POST /conditional-route', async () => {
    const log = jest.spyOn(global.console, 'log');
    await request(app).post("/conditional-route").send({ third: true });
    await awaitConsoleLog(300, log.mock.calls)
    expect(log.mock.calls).toMatchSnapshot();
  })
})