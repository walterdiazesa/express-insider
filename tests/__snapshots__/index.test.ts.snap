// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trail on custom configuration Should NOT output GET /no-response-route payload 1`] = `
[
  [
    "[test-id] [32mGET /no-response-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [32mGET /no-response-route [33m0 ms [32m200 OK[0m",
  ],
  [
    "[test-id] [32mGET /no-response-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 1] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)>[0m",
  ],
  [
    "[test-id]                                             (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 3] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)>[0m",
  ],
  [
    "[test-id]                                             (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 4] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)>[0m",
  ],
  [
    "[test-id]                                        (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                        [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)> [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32mlastmiddle [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 5] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)>[0m",
  ],
  [
    "[test-id]                                             (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)> [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32mlastmiddle [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 6] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)>[0m",
  ],
  [
    "[test-id]                                             (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)> [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 7] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)> [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)>[0m",
  ],
  [
    "[test-id]                                             (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 8] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)>[0m",
  ],
  [
    "[test-id]                                             (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)> [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 9] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)> [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)>[0m",
  ],
  [
    "[test-id]                                             (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 10] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)> [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)>[0m",
  ],
  [
    "[test-id]                                             (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32mlastmiddle [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 11] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)>[0m",
  ],
  [
    "[test-id]                                        (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                        [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (cleanup) [33m0 ms[0m",
  ],
  [
    "[test-id]                                             (total) [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)> [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32mlastmiddle [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should handle [CASE 12] with POST /conditional-route 1`] = `
[
  [
    "[test-id] [33mPOST /conditional-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [33mPOST /conditional-route[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (1)> [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32m<anonymous (0)> [33m0 ms[0m",
  ],
  [
    "[test-id]                              [32mlastmiddle[0m",
  ],
  [
    "[test-id]                                        (send response) [33m0 ms[0m",
  ],
  [
    "[test-id]                                        [32m200 OK[0m, total elapsed time: [33m0 ms[0m",
  ],
  [
    "[test-id] [33mPOST /conditional-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;

exports[`trail on custom configuration Should output GET /response-route payload 1`] = `
[
  [
    "[test-id] [32mGET /response-route [0m[1mstart[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mquery [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mexpressInit [33m0 ms[0m",
  ],
  [
    "[test-id] [36mMiddleware [32mjsonParser [33m0 ms[0m",
  ],
  [
    "[test-id] [35mRoute [32mGET /response-route [33m0 ms [32m200 OK[0m",
  ],
  [
    "[test-id] [1m[RESPONSE]:[0m {
            [33mpayload: [0m[0m[32m"some payload"[0m,
            [33mitems: [0m[
              {
                [33mid: [0m[0m[34m1[0m
              },
              {
                [33mid: [0m[0m[34m2[0m
              }
            ]
          }",
  ],
  [
    "[test-id] [32mGET /response-route [0m[1mfinish[0m, elapsed time since begin: [33m0 ms[0m",
  ],
]
`;
